### PR TITLE
Trigger completion in 'new()' expression

### LINKED ISSUE #46397: Object initializer completion doesn't trigger on target-typed new
I'm on latest VS 2019 preview, and I've noticed auto-complete popup doesn't work for new() class initializers. As in:
```
var color = new Color()
{
  // ctrl+space and suggestions are "r, g, b, a"
}

Color color = new()
{
  // ctrl+space has no response
}
```

Reported [here](https://github.com/dotnet/roslyn/issues/28489#issuecomment-665205328) (thanks @SirIntruder)

### LINKED ISSUE #47381: Brace completion on target-typed-new introduces a new line
I intend to type `C c = new() { ... };` but a newline gets introduced, which is pretty annoying.

target-typed-new-completion

Relates to https://github.com/dotnet/roslyn/issues/46397

### PR DESCRIPTION
**Pull request by alrz**
Fixes #46397
Fixes #47381

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectAndWithInitializerCompletionProvider.cs [hunk 1]
```diff
@@ -144,6 +145,18 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
...
             }
 
+            // new() { bar = $$
+            if (token.Parent.Parent.IsKind(SyntaxKind.ImplicitObjectCreationExpression, out ImplicitObjectCreationExpressionSyntax implicitObjectCreation))
+            {
+                var type = semanticModel.GetTypeInfo(implicitObjectCreation, cancellationToken).ConvertedType;
+                if (type is ITypeParameterSymbol typeParameterSymbol)
+                {
+                    return Tuple.Create<ITypeSymbol, Location>(typeParameterSymbol.GetNamedTypeSymbolConstraint(), token.GetLocation());
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectAndWithInitializerCompletionProvider.cs:155]: this feels duplicative.  i'd prefer we jsut return the type, and have a higher layer say what to do in teh case of a type-parameter.  (this applies to the ObjCreation case above as well).

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectAndWithInitializerCompletionProvider.cs:155]: thanks! :)

**Comment by alrz** [comment]
ping @CyrusNajmabadi 

I think this won't make it to 16.8?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 2]
```diff
@@ -211,6 +211,8 @@ public static (SyntaxToken openBrace, SyntaxToken closeBrace) GetBraces(this Syn
                     return (property.OpenBraceToken, property.CloseBraceToken);
                 case WithExpressionSyntax withExpr:
                     return (withExpr.Initializer.OpenBraceToken, withExpr.Initializer.CloseBraceToken);
+                case ImplicitObjectCreationExpressionSyntax newExpr when newExpr.Initializer != null:
+                    return (newExpr.Initializer.OpenBraceToken, newExpr.Initializer.CloseBraceToken);
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxNodeExtensions.cs:215]:
```suggestion
                case ImplicitObjectCreationExpressionSyntax { Initializer: { } initializer }:
                    return (initializer.OpenBraceToken, initializer.CloseBraceToken);
```

Comment (alrz) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxNodeExtensions.cs:215]: that wont work due to a bug https://github.com/dotnet/roslyn/issues/44534

Comment (alrz) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxNodeExtensions.cs:215]: wait, it doesnt only if we use newExpr. I can do that

**Comment by alrz** [comment]
@sharwell do you have any other comments?

**Comment by CyrusNajmabadi** [comment]
Ok vacation. Can wait until next week, or if @sahrwell can look this week is could go forward.

**Review by genlu — APPROVED** [review]

**Review by genlu — APPROVED** [review]
LGTM. Thanks!

**Review by sharwell — APPROVED** [review]

**Pull request merged**